### PR TITLE
Make iOS export fast and easy

### PR DESCRIPTION
**Pull request by samgreen**
I'd like to get to 1-click deploy from the editor for iOS games. The first step is ensuring the export pipeline is solid, up-to-date and has very reasonable defaults.

# Changes
* Move to **Truly Automatic** ™️ code-signing. The trick is providing a default value of `iPhone Developer` and `iPhone Distribution` to pick from available code signing identities (Beward, this can cause #24866 and loss of sanity). The old behavior is still here if you choose to modify the export settings. Most of this process should now be automatic for _most users_
* Depend on modifying the `.xcodeproj` **much** less. We want to stay away from modifying this file and instead let Xcode manage it. To that end, most settings have been moved to `xcconfig` files.
* Add `Common.xcconfig` these settings are applied to all (Debug and Release) builds
* Generate tweakable settings in `godot.xcconfig` and add to `xcodeproj`
* Don't require any icons for successful export, iOS will use a white icon with concentric circles until the art is in place. This was just _one more_ hurdle people needed to see their games run on their devices, I've personally spent a bit of time generating random PNGs just to skip an error in this part of the export. Note that we (obviously) still set the icons if the developer provides them.
* Disambiguate the two versions needed on export. One is a 3 digit version that MUST increment, one is for public display.
* Remove **Access WiFi** capability from export option, as it does [something else](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_networking_wifi-info?language=objc).
* Bump deployment target to iOS 10 (to match the export templates).
* Enabled just about every optional warning and set some to errors, to ensure the best hygiene. Also fixed all warnings that resulted from this.

# Resolves
* #25128 - .xcconfig files autoescape all strings by default
* #24866 - This was due to the project not actually forcing the placeholder if the user typed nothing for the code signing identities
* #22197 - Ensures x86_64 slices is added to lipo fat binary
* #19996 - Use injected entitlements based on provisioning profile
* #15997 - Only 2 values (team ID and bundle ID) are now required to properly export to iOS, all other settings are tweakable after export via the `.xcconfig` file.

# Clean Up
* Remove `CFBundleSignature` from `plist` (This is an OS X key).
* Remove `CFBundleIcons` from `plist (We previously migrated to Asset Catalogs)
* Resolve all build warnings from exported Xcode project

# Proposal for future work
* Remove identity and export method entirely and depend on Xcode export from Archive (this allows re-exporting the same build repeatedly with different certs, profiles, and capabilities).
* Add better documentation around `capabilities` and what they mean.

### CODE REVIEW

**Comment by samgreen** [comment]
Example of the final `godot.xcconfig` that developers will see in Xcode:

```
#include "Common.xcconfig"

//
//  Build Settings
//
ENABLE_BITCODE = NO;

//
//  Deploy Settings
//
TARGETED_DEVICE_FAMILY = 1,2; // Target iPhone and iPad
IPHONEOS_DEPLOYMENT_TARGET = 10.0;

ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
ASSETCATALOG_COMPILER_LAUNCHIMAGE_NAME = LaunchImage;

PRODUCT_NAME = $(TARGET_NAME);

// 
// Godot Exported from Editor
//
INFOPLIST_FILE = Prickles/Prickles-Info.plist;
PRODUCT_BUNDLE_IDENTIFIER = com.samgreen.prickles;
DEVELOPMENT_TEAM = USYSZ347Y9;
ARCHS = arm64;
FRAMEWORK_SEARCH_PATHS = $(inherited) Prickles;
OTHER_LDFLAGS = $(inherited);
```

File: misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -197,7 +199,7 @@
 						ProvisioningStyle = Automatic;
 						SystemCapabilities = {
 							com.apple.AccessWiFi = {
-								enabled = $access_wifi;
+								enabled = 0;
```
Comment (samgreen) [inline misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: All games can still access wifi. This is poorly named (on Apple's part).

File: platform/iphone/export/export.cpp [hunk 2]
```diff
@@ -300,61 +298,95 @@ void EditorExportPlatformIOS::get_export_options(List<ExportOption> *r_options)
...
+	// Pass linker flags
+	strnew += "OTHER_LDFLAGS = $(inherited)";
+	if (p_config.linker_flags.length() != 0) {
+		strnew += " " + p_config.linker_flags;
+	}
+	strnew += ";\n";
+
+	// if enabled, add development (sandbox) APNS entitlements
+	if ((bool)p_preset->get("capabilities/push_notifications")) {
+		strnew += "\n\"CODE_SIGN_ENTITLEMENTS[sdk=iphoneos*]\" = " + binary_path + ".entitlements;\n";
```
Comment (samgreen) [inline platform/iphone/export/export.cpp]: We can probably drop this in the future and let Xcode generate entitlements as well what it's already doing in this PR.

**Comment by samgreen** [comment]
@akien-mga I just realized there's a section in the docs about cross compiling iOS apps on Linux. I'm not sure if this is still supported, or even a goal of this project. Certainly the code signing aspects wouldn't work. This PR will further add a dependence on Xcode, but that's already *required* for submitting to the App Store.

**Comment by akien-mga** [comment]
As long as we can still *build* the templates themselves on Linux (which we currently do for official templates), that should be fine :)

File: platform/iphone/export/export.cpp [hunk 3]
```diff
@@ -243,26 +244,24 @@ void EditorExportPlatformIOS::get_export_options(List<ExportOption> *r_options)
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "custom_package/debug", PROPERTY_HINT_GLOBAL_FILE, "*.zip"), ""));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "custom_package/release", PROPERTY_HINT_GLOBAL_FILE, "*.zip"), ""));
 
-	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/app_store_team_id"), ""));
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/app_store_team_id", PROPERTY_HINT_PLACEHOLDER_TEXT, "ZXSYZ493U4"), ""));
```
Comment (alcroito) [inline platform/iphone/export/export.cpp]: Is ZXSYZ493U4 just a dummy value?

File: misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -100,7 +100,7 @@
 		D0BCFE2B18AEBDA2004A7AAE = {
 			isa = PBXGroup;
 			children = (
-				87D30C9722235A2200E688DC /* Config.xcconfig */
+				87D30C9722235A2200E688DC /* Config.xcconfig */,
```
Comment (akien-mga) [inline misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: Would be nice if you could squash this commit into its parent (40a2122b).

Comment (akien-mga) [reply misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: and/or group some of the other related commits that iteratively modify the same file.

**Comment by akien-mga** [comment]
I'll tentatively put this on the 3.1 milestone even though it's quite late in the process, as the iOS export pipeline is far from optimal currently, so those changes might be very welcome in our new minor release.

I'll need help from experienced iOS devs to validate these changes ASAP though, cc @bruvzg @BastiaanOlij @endragor.

File: misc/dist/ios_xcode/godot_ios/Common.xcconfig [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+// This setting is on for new projects as of Xcode ~6.3, though it is still not
+// the default. It warns if the same variable is declared in two binaries that
+// are linked together.
+GCC_NO_COMMON_BLOCKS = YES
+
+// This warnings detects when a function will recursively call itself on every
+// code path though that function. More information can be found here:
+// http://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20131216/096004.html
+CLANG_WARN_INFINITE_RECURSION = YES
```
Comment (akien-mga) [inline misc/dist/ios_xcode/godot_ios/Common.xcconfig]: Missing newline at EOF.

File: misc/dist/ios_xcode/godot_ios/godot.xcconfig [hunk 6]
```diff
@@ -19,3 +18,5 @@ WRAPPER_EXTENSION = app;
 
 ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 ASSETCATALOG_COMPILER_LAUNCHIMAGE_NAME = LaunchImage;
+
+PRODUCT_NAME = "$(TARGET_NAME)";
```
Comment (akien-mga) [inline misc/dist/ios_xcode/godot_ios/godot.xcconfig]: Missing newline at EOF.

File: misc/dist/ios_xcode/godot_ios/godot.xcconfig [hunk 7]
```diff
@@ -6,7 +6,7 @@
 CONFIGURATION_BUILD_DIR = "$(BUILD_DIR)/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)";
 SDKROOT = iphoneos;
 VALID_ARCHS = "armv7 armv7s arm64 x86_64";
-ENABLE_BITCODE = NO
+ENABLE_BITCODE = NO;
```
Comment (akien-mga) [inline misc/dist/ios_xcode/godot_ios/godot.xcconfig]: Would be good if this hotfix was amended into the commit where the line was added in the first place.

File: misc/dist/ios_xcode/godot_ios/godot.xcconfig [hunk 8]
```diff
@@ -4,7 +4,7 @@
 //  Build Settings
 //
 SDKROOT = iphoneos;
-VALID_ARCHS = "armv7 armv7s arm64 x86_64";
+VALID_ARCHS = armv7 armv7s arm64 x86_64;
```
Comment (akien-mga) [inline misc/dist/ios_xcode/godot_ios/godot.xcconfig]: Same for this one.

File: platform/iphone/export/export.cpp [hunk 9]
```diff
@@ -244,20 +244,20 @@ void EditorExportPlatformIOS::get_export_options(List<ExportOption> *r_options)
...
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/code_sign_identity_release", PROPERTY_HINT_PLACEHOLDER_TEXT, "iPhone Distribution"), "iPhone Distribution"));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::INT, "application/export_method_release", PROPERTY_HINT_ENUM, "App Store,Development,Ad-Hoc,Enterprise"), 0));
 
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/name", PROPERTY_HINT_PLACEHOLDER_TEXT, "Game Name"), ""));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/info"), "Made with Godot Engine"));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/identifier", PROPERTY_HINT_PLACEHOLDER_TEXT, "com.example.game"), ""));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/public_version"), "1.0.0"));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/version"), "1.0.0"));
-	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/copyright"), ""));
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/copyright"), "2019"));
```
Comment (akien-mga) [inline platform/iphone/export/export.cpp]: Is `application/copyright` meant to be a year?

File: platform/iphone/export/export.cpp [hunk 10]
```diff
@@ -257,7 +257,7 @@ void EditorExportPlatformIOS::get_export_options(List<ExportOption> *r_options)
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/identifier", PROPERTY_HINT_PLACEHOLDER_TEXT, "com.example.game"), ""));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/public_version"), "1.0.0"));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/version"), "1.0.0"));
-	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/copyright"), "2019"));
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "application/copyright"), ""));
```
Comment (akien-mga) [inline platform/iphone/export/export.cpp]: This answers my previous question... should be squashed in the commit where it was changed.

**Comment by akien-mga** [comment]
The changes look good to me overall, but I'd like the git history to be amended a bit to make it easier to follow the changes, removing intermediate changes that are no longer relevant once merged.

So typo fixes and cleanups should be made directly in the original commit that added the need for such fixes. The final commits that we do merge should not add bugs, even temporarily; each commit should bring the engine from one valid state to another valid state, so that changes can be bisected if need be.

**Comment by hpvb** [comment]
I'd very much like to push this to 3.2, maybe it can go in 3.1.1 if we have to. This will take a while to sort out on my end.

**Comment by samgreen** [comment]
I agree that we shouldn’t rush this out. This is a vast change that deserves more testing.

File: misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj (hunk 1 above)
Comment (BastiaanOlij) [reply misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: yeah Apple has some strange ideas about these capabilities. Unless someone runs into a situation where this needs to be turned on, take it out.

File: misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -333,132 +335,51 @@
...
-				INFOPLIST_FILE = "$binary/$binary-Info.plist";
-				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
-				LIBRARY_SEARCH_PATHS = (
-					"$(inherited)",
-				);
-				PRODUCT_BUNDLE_IDENTIFIER = $identifier;
-				PRODUCT_NAME = "$(TARGET_NAME)";
-				PROVISIONING_PROFILE = "$provisioning_profile_uuid_debug";
-				TARGETED_DEVICE_FAMILY = "1,2";
-				VALID_ARCHS = "armv7 armv7s arm64 i386 x86_64";
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: I think its worth discussing whether we want to get rid of the 32bit stuff as well. Apple is about to pull the plug on that anyway. Or maybe wait for 3.2 for that one?

File: platform/iphone/export/export.cpp (hunk 3 above)
Comment (BastiaanOlij) [reply platform/iphone/export/export.cpp]: It's probably Sams team id so we can now all release games in his name :wink:

File: platform/iphone/export/export.cpp (hunk 2 above)
Comment (BastiaanOlij) [reply platform/iphone/export/export.cpp]: I'm a fan of letting XCode do as much of the work for us as possible, that way if apple decides to change/enhance it we should end up piggy backing on that change.

File: misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj (hunk 11 above)
[the remark this answers is not in the document]
Comment (akien-mga) [reply misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj]: I'd wait for 3.2, but then yeah, we can likely drop 32-bit stuff. See discussion in godotengine/godot-docs#2212.

**Comment by BastiaanOlij** [comment]
Looks pretty good to me at first glance, once it's rebased (and maybe squashed) I can give it a twirl next time I'm working on ARKit and see if the changes work for me. 

I also second moving this to 3.2, that way we can spend time perfecting this even if that means merging it soon after 3.1 is released and then doing additional fixes as we push this further along and add more or the settings, dialog strings, etc. 

The current export in master is functional with a little bit of setup on the XCode side and this may be too big a change to risk pushing the release of 3.1 out. 

One thing I would like to see added is more default on the settings so that really all that is needed to test a build is to put in your team ID (or even leave that empty and do that as a last step when opening up xcode, not everyone will know for sure what to enter here).
An example of this are the splash screens you must supply in all different sizes. Even if these are blank images or simple images with the Godot logo on it, that just allows you to quickly test you game without having to make any. Those you would generally not make until you're nearing the end of your development cycle so saving that time at the start would help.

**Comment by reduz** [comment]
looks great, can't wait for it to be merged

**Comment by PabloNeirotti** [comment]
Hey! I am really looking forward to see this out there. I currently can't export to iOS on the latest Godot version. On Xcode I get the error "The project 'project name here' is damaged and cannot be opened due to a parse error." (and yes, I filled in all the fields with information I compiled across multiple issues).

If you need help testing this please let me know.

As from a usability standpoint, I agree parameters should be pre-filled out to output a project that requires the least amount of configuration to get to test on a real device. Even the icon (just put in Godot icons ;P). I can provide those if you need it.

**Comment by akien-mga** [comment]
For the reference, the changes in this PR seem great overall, but the git history needs some work as mentioned above. More importantly, #22784 should likely be merged and tested first before we tackle the export workflow.

**Comment by zacharytyhacz** [comment]
@samgreen  @akien-mga  Any update on this PR? future addition to 3.2? I've tried many solutions regarding the issues referenced in this PR that will, or may,  be fixed and still having trouble exporting to IOS for development on development Apple account. 

Thanks and nice job Sam

**Comment by tadaaagames** [comment]
Is this going to be in 3.2? Very much looking forward to it!

**Comment by akien-mga** [comment]
Moving to 4.0 milestone as this is not ready to merge.

**Comment by akien-mga** [comment]
See also #33086 for a related PR with probably a smaller scope.

File: platform/iphone/export/export.cpp (hunk 9 above)
Comment (samgreen) [reply platform/iphone/export/export.cpp]: Can be the company name and year something like "Godot ©2019"

**Comment by samgreen** [comment]
@akien-mga @godotengine/ios 

This has been rebased and updated.

**Comment by aaronfranke** [comment]
@samgreen What is the status of this PR? The two PRs linked by Akien are both closed. If this is still desired, this should be rebased and re-tested so that we know it still works.

**Comment by samgreen** [comment]
I don’t have a good answer here. I’m happy to make this and any further changes. I’d love to see iOS export get some love. This is a huge barrier to wider adoption.

**Comment by trojanfoe** [comment]
@samgreen Hi, I am kinda new to Godot, but not to Xcode.  One thing I don't understand is why the Godot "runtime" isn't exported to Xcode as source instead of built binary, which would allow Xcode to build the source for device or simulator?  This would presumably also cut down on the maintenance effort of the exporter itself?

**Comment by trojanfoe** [comment]
@samgreen 
> There is no Xcode on Linux.

No what I meant was the iOS exporter is comprised of the source files of the Godot runtime and the necessary Xcode project files so when the user exports to iOS on their Mac they then built it to whatever target they like (device, simulator, arm64, x64, whatever).  Has this not been considered?

**Comment by clofresh** [comment]
I can test this change but I'm having trouble rebasing onto master or 3.x as there's a bunch of conflicts in godot_ios.xcodeproj/project.pbxproj that I don't understand. If someone resolves them, I can test it on my end.

**Comment by Vennnot** [comment]
Seems that if this was retested we could see if we merge it or not

**Comment by akien-mga** [comment]
This PR is sadly abandoned, but https://github.com/godotengine/godot/pull/70662 should supersede it.

**Pull request closed without merging**